qiskit/providers/backend.py

```diff
@@ -219,3 +224,243 @@ class can handle either situation.
+    will build a :class:`~qiskit.providers.models.BackendConfiguration` object
+    and :class:`~qiskit.providers.models.BackendProperties` from the attributes
+    defined in this class for backwards compatibility.
+    """
+
+    version = 2
+
+    def __init__(self, provider, **fields):
+        self._options = self._default_options()
+        self._provider = provider
+        self._coupling_map = None
+        if fields:
+            for field in fields:
+                if field not in self._options.data:
+                    raise AttributeError("Options field %s is not valid for this backend" % field)
+            self._options.update_config(**fields)
+
+    @property
+    def gates(self) -> List[Gate]:
+        """A list of Gate instances that the backend supports."""
+        return list(self.gate_map.gates)
+
+    @property
+    def gate_names(self) -> List[str]:
+        """A list of Gate names that the backend supports."""
+        return list(self.gate_map.gate_names)
+
+    @property
+    @abstractmethod
+    def gate_map(self):
+        """A :class:`qiskit.transpiler.GateMap` object for the backend."""
+        pass
+
+    @property
+    @abstractmethod
+    def num_qubits(self) -> int:
+        """Return the number of qubits the backend has."""
+        pass
+
+    @property
+    @abstractmethod
+    def supported_instruction(self) -> List[Instruction]:
+        """Return the list of supported non-gate instructions for the
+        backend.
+
+        The expectation this will be things like Measure, Reset, Save, etc.
+        """
+        pass
+
+    @property
+    def coupling_map(self):
+        """Return the :class:`~qiskit.transpiler.CouplingMap` object"""
+        return self.gate_map.coupling_map()
+
+    @property
+    @abstractmethod
+    def conditional(self) -> bool:
+        """Return bool whether the target can execute gates with classical
+        conditions."""
+        pass
+
+    @property
+    @abstractmethod
+    def max_shots(self) -> int:
+        """Return the maximum number of shots supported by the backend.
+
+        If there is no limit this will return None
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def max_experiments(self):
+        """The maximum number of circuits (or Pulse schedules) that can be
+        run in a single job.
+
+        If there is no limit this will return None
+        """
+        pass
+
+    @classmethod
+    @abstractmethod
+    def _default_options(cls):
+        """Return the default options
+
+        This method will return a :class:`qiskit.providers.Options`
+        subclass object that will be used for the default options. These
+        should be the default parameters to use for the options of the
+        backend.
+
+        Returns:
+            qiskit.providers.Options: A options object with
+                default values set
+        """
+        pass
+
+    def t1(self, qubit: int):
```

Comment:
So are these not part of the `properties` anymore?  Also if we are adding qubit-based properties doesn't single-qubit gate errors go here?

Reply:
Yeah, for backendv2 I'm working on making things flatter and more user accessible. The problem we've had with the backend/provider interface since it was first introduced is it's not really designed as a user facing interface it's been basically a python object model wrapping the iqx api response formats (this is why we've had horrible nested structures etc because it matches the json payload format, it dates back to the marshmallow usage). For backend V2 (and what I really wanted to do in BackendV1 but deferred for backwards compat/ease of migration) I'm hoping to make the interface a native class interface that is easier for users to interact with and query backend parameters they need but also for non-ibm providers to start to provide more of this data. Also this should hopefully make it clear the difference between paramters that are required, optional, and unique to a specific provider. Right now I don't think any providers use `BackendProperties` (and it's hard to know how) except for ibmq. I also am making sure to leave `configuration()` and `properties()` out of the abstract class definition though, so existing providers can add them for backwards compatibility.

The single qubit gate errors (and duration) are bundled in the gate map data structure (I might change the name) which is a data structure for backends to provide a target for the transpiler which combines the gates, args, and any gate properties: https://github.com/Qiskit/qiskit-terra/pull/5885/files#diff-4b1645b85aea633298ee64a56a5f6bd55c3028df52fbb947fe8de12be23b29dfR43

Reply:
I am def in favor of redoing how this is handled.  e.g see https://github.com/Qiskit-Partners/qiskit-ibm/issues/85.

I am curious how this all works though.  Namely, for example, we have two different ways of doing readout on newer IBM systems. the trad. way and ESP.  There are therefore two values for `readout_error` that need to be supported.  Of course, other vendors backends may not have these options.  So how to unify all this seems tricky.

Reply:
Well this is still WIP, so we can improve the structure here as it makes sense. I forgot about the multiple readout types and didn't that factor into the current version. How I'm thinking we should treat is to rename `GateMap` to `Target` and have it be for instructions and gates, then the ibm provider can populate a backend's target object with 2 entries like:

```python
        from qiskit.transpiler import Target
        from qiskit.circuit.library import UGate, CXGate, 
        from qiskit.circuit import Parameter
        gmap = Target()
        theta = Parameter('theta')
        phi = Parameter('phi')
        lam = Parameter('lambda')
        u_props = {
            (0,): GateProperties(length=5.23e-8, error=0.00038115),
            (1,): GateProperties(length=4.52e-8, error=0.00032115),
        }
        gmap.add_gate(UGate(theta, phi, lam), [(0,), (1,)], properties=u_props)
        cx_props = {
            (0,1): GateProperties(length=5.23e-7, error=0.00098115),
            (1,0): GateProperties(length=4.52e-7, error=0.00132115),
        }
        gmap.add_gate(CXGate(), [(0, 1), (1, 0)], properties=cx_props)
        measure_props = {
            (0,): GateProperties(length=4e-8, error = 0.0002),
            (1,): GateProperties(length=4e-8, error = 0.0002),
        }
        gmap.add_gate(Measure(),  [(0,), (1,)], properties=measure_props)
        
        esp_readout_props = {
            (0,): GateProperties(length=2e-8, error = 0.0001),
            (1,): GateProperties(length=2e-8, error = 0.00012),
        }
        # esp readout behaves like a normal measure but has a unique name to distinguish it from a normal measurement
        gmap.add_gate(Measure(),  [(0,), (1,)], properties=measure_props, name=esp_readout)
```

Then the target has an entry for both measure and esp_readout each with different duration and error rates and the transpiler will have the necessary information it needs about both instructions.

Reply:
Ok in https://github.com/Qiskit/qiskit-terra/pull/5885/commits/e61abcd30ff2bf29639b55a69b9b5e0230069484 I dropped the optional `readout_error` and `readout_length` methods and renamed `GateMap` target and made it for all operations (not just gates). This will let us represent different parameters for normal measurement and esp readout by adding two entries to the target for each type of measurement.